src/operations/client_bulk_write/executor.ts

```diff
@@ -65,15 +71,42 @@ export class ClientBulkWriteExecutor {
     } else {
       const resultsMerger = new ClientBulkWriteResultsMerger(this.options);
       // For each command will will create and exhaust a cursor for the results.
-      let currentBatchOffset = 0;
       while (commandBuilder.hasNextBatch()) {
         const cursor = new ClientBulkWriteCursor(this.client, commandBuilder, this.options);
-        const docs = await cursor.toArray();
-        const operations = cursor.operations;
-        resultsMerger.merge(currentBatchOffset, operations, cursor.response, docs);
-        // Set the new batch index so we can back back to the index in the original models.
-        currentBatchOffset += operations.length;
+        try {
+          await resultsMerger.merge(cursor);
+        } catch (error) {
+          // Write concern errors are recorded in the writeConcernErrors field on MongoClientBulkWriteError.
+          // When a write concern error is encountered, it should not terminate execution of the bulk write
+          // for either ordered or unordered bulk writes. However, drivers MUST throw an exception at the end
+          // of execution if any write concern errors were observed.
+          if (error instanceof MongoServerError && !(error instanceof MongoClientBulkWriteError)) {
+            // Server side errors need to be wrapped inside a MongoClientBulkWriteError, where the root
+            // cause is the error property and a partial result is to be included.
+            const bulkWriteError = new MongoClientBulkWriteError({
```

Comment:
Thoughts on using `cause` here instead of the `error` property? https://mongodb.slack.com/archives/C035ZJL6CQN/p1727975432135529

Reply:
Updated to use cause, since it's idiomatic JS.